src/examples/AppUartModule.cpp

```diff
@@ -0,0 +1,270 @@
+////////////////////////////////////////////////////////////////////////////////
+// /****************************************************************************
+// **
+// ** Copyright (C) 2015-2021 M-Way Solutions GmbH
+// ** Contact: https://www.blureange.io/licensing
+// **
+// ** This file is part of the Bluerange/FruityMesh implementation
+// **
+// ** $BR_BEGIN_LICENSE:GPL-EXCEPT$
+// ** Commercial License Usage
+// ** Licensees holding valid commercial Bluerange licenses may use this file in
+// ** accordance with the commercial license agreement provided with the
+// ** Software or, alternatively, in accordance with the terms contained in
+// ** a written agreement between them and M-Way Solutions GmbH.
+// ** For licensing terms and conditions see https://www.bluerange.io/terms-conditions. For further
+// ** information use the contact form at https://www.bluerange.io/contact.
+// **
+// ** GNU General Public License Usage
+// ** Alternatively, this file may be used under the terms of the GNU
+// ** General Public License version 3 as published by the Free Software
+// ** Foundation with exceptions as appearing in the file LICENSE.GPL3-EXCEPT
+// ** included in the packaging of this file. Please review the following
+// ** information to ensure the GNU General Public License requirements will
+// ** be met: https://www.gnu.org/licenses/gpl-3.0.html.
+// **
+// ** $BR_END_LICENSE$
+// **
+// ****************************************************************************/
+////////////////////////////////////////////////////////////////////////////////
+
+#include <AppUartModule.h>
+#include <GlobalState.h>
+#include <Logger.h>
+#include <Node.h>
+#include <Utility.h>
+#include <mini-printf.h>
+
+AppUartModule::AppUartModule()
+    : Module(APP_UART_MODULE_ID, "appuart")
+    , logQueue(logBuffer, APP_UART_LOG_BUFFER_SIZE)
+{
+    // Register callbacks n' stuff
+
+    // Enable the logtag for our vendor module template
+    GS->logger.EnableTag("APPUART");
+    // If MACONN is active, the log will be output on the packet transmission itself, so the application log will be
+    // sent forever.
+#if IS_ACTIVE(APP_UART)
+    GS->logger.DisableTag("MACONN");
+#endif
```

Comment:
With the existence of `FruityHal`, I think that I should avoid calling the libraries provided by nRF SDK directly in `AppUartModule` in order to use critical section and atomic variables.
Is it OK to define functions that use critical section and atomic in `FruityHal`, and implement them in `FruityHalNrf` using `CRITICAL_REGION_ENTER()` and [Atomic operations API](https://infocenter.nordicsemi.com/index.jsp?topic=%2Fcom.nordic.infocenter.sdk5.v15.3.0%2Fgroup__app__util__platform.html)?

If I don't use the library, I think I need to use inline assembly. I've never used assembly before, so it will take some time to implement. (If you have any advice on how to implement this without using assembly, I would appreciate it.)

Reply:
First off all I'm very sorry for taking so long to write back.

You don't have to use the Nordic headers. For atomic variables you can generally use C++'s `<atomic>` header. The ARM GCC (4.9) we use for compiling includes this header in it's freestanding implementation.

I'll put some pointers together and let you know, likely sometime tomorrow!

Reply:
I have completed the correction for the content of Part 1 using atomic variables. I should have done a little more research on the use of atomic variables myself. Thank you for your kind advice.

The content of Part 2 has also been fixed by saving the `node ID` of the connection target after establishing the connection and using the connection handle.
The problem of `AppUartModule` sending and receiving messages to `node ID` within the scope of the `APP` has also been fixed.